include/mqtt/property_parse.hpp

```diff
@@ -17,273 +17,229 @@ namespace mqtt {
 namespace v5 {
 namespace property {
 
-template <typename It>
 inline
-mqtt::optional<property_variant> parse_one(It& begin, It end) {
-    if (begin == end) return mqtt::nullopt;
+std::pair<mqtt::optional<property_variant>, buffer> parse_one(buffer buf) {
+    if (buf.empty()) return { mqtt::nullopt, std::move(buf) };
     try {
-        auto it = begin;
-        switch (static_cast<property::id>(*it++)) {
+        auto id = static_cast<property::id>(buf.front());
+        buf = buf.substr(1);
```

Comment:
Careful with calls to substr, this causes an atomic increment in the shared pointer, and then an atomic decrement.

Reply:
Yes, I should avoid unnecessary atomic increment and decrement via `share_ptr_array`'s copy constructor.

I come up with https://wandbox.org/permlink/UramwEowhKy1oRFw. This could be another option.

I didn't check all locations I call `substr()` yet. If `mqtt::string_view` is obviously easy and clear, I use it.

Reply:
Using rvalue references is a good way to avoid the problem in the generic sense.

But I think that there is still room for improvement in the places where MQTT::buffer is used, as well.

Reply:
All `buf = buf.substr(N)` (buf : buffer) are replaced with `buf = std::move(buf).substr()` for the first step.